src/internal/clusterstate/v2.8.0/pfsdb.go

```diff
@@ -0,0 +1,137 @@
+package v2_8_0
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/jmoiron/sqlx"
+	"google.golang.org/protobuf/proto"
+
+	v2_7_0 "github.com/pachyderm/pachyderm/v2/src/internal/clusterstate/v2.7.0"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/pfs"
+)
+
+const (
+	// ReposChannelName
+	ReposChannelName = "pfs_repos"
+)
+
+func generateTriggerFunctionStatement(schema, table, channel string) string {
+	template := `
+	CREATE OR REPLACE FUNCTION %s.notify_%s() RETURNS TRIGGER AS $$
+	DECLARE
+		row record;
+		payload text;
+	BEGIN
+		IF TG_OP = 'DELETE' THEN
+			row := OLD;
+		ELSE
+			row := NEW;
+		END IF;
+		payload := TG_OP || ' ' || row.id::text;
+		PERFORM pg_notify('%s', payload);
+		return row;
+	END;
+	$$ LANGUAGE plpgsql;
+
+	CREATE TRIGGER notify
+		AFTER INSERT OR UPDATE OR DELETE ON %s.%s
+		FOR EACH ROW EXECUTE PROCEDURE %s.notify_%s();
+	`
+	return fmt.Sprintf(template, schema, table, channel, schema, table, schema, table)
+}
+
+func ListReposFromCollection(ctx context.Context, q sqlx.QueryerContext) ([]Repo, error) {
```

Comment:
can this be private?

Reply:
I actually use this outside of this package in `src/internal/clusterstate/v2.8.0_test.go`. I toiled with the idea of whether to test this stuff in the v2.8.0 package or in the general clusterstate package. Ultimately, I decided that clusterstate tests should be in the broader clusterstate package instead of in the specific versions of each. This is because we want to test state transitions from one state to the next and create scenarios that span multiple states potentially.